### PR TITLE
Avoid unnecessary monomorphization of inline asm related functions

### PR DESCRIPTION
**Pull request by bjorn3**
This should reduce build time for codegen backends by avoiding duplicated monomorphization of certain inline asm related functions for each passed in closure type.

### CODE REVIEW

File: compiler/rustc_passes/src/intrinsicck.rs [hunk 1]
```diff
@@ -413,7 +411,7 @@ impl<'tcx> ExprVisitor<'tcx> {
                             let msg = format!(
                                 "register class `{}` requires at least one of the following target features: {}",
                                 reg_class.name(),
-                                features.join(", ")
+                                features.iter().map(|f| f.as_str()).collect::<Vec<_>>().join(", ")
```
Comment (nagisa) [inline compiler/rustc_passes/src/intrinsicck.rs:414]: Why is this necessary? The former code is definitely more readable. A good way to avoid an intermediate allocation here would be to write:

```
features.iter().map(...).intersperse(", ").collect::<String>()
```

**Review by nagisa — APPROVED** [review]
Overall LGTM.

Feel free to r=me after fixing the nit or not, whichever you prefer.

File: compiler/rustc_passes/src/intrinsicck.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_passes/src/intrinsicck.rs]: TIL

**Comment by bjorn3** [comment]
No perf changes and fixed the nit.

@bors r=nagisa

**Comment by bjorn3** [comment]
Rebased. The only conflict was in the list of pre-defined symbols.

@bors r=nagisa

**Comment by bjorn3** [comment]
This saved about 200KB of 11MB (2%) for librustc_codegen_cranelift.so. It looks like it also saved a second or two of the cg_clif build time, but that one is pretty noisy anyway.
